### PR TITLE
'make proto' now requires a <1 month old docker image

### PR DESCRIPTION
**Pull request by msteffen**
This is a bit different than what I'd originally proposed (locking the version of the protobuf compiler that `make proto` uses) but I went with this approach because:
1. My curl request to get the latest commit in e.g. `github.com/golang/protobuf` got rate-limited almost immediately. To get a higher rate limit I would've had to put some kind of auth credentials in the `curl` command, and as this script is checked in to our public repo, I wasn't comfortable putting auth creds in it (even an OAuth token that can't do anything could be used by others and count against our rate limit) and having one credential used by whole company didn't seem scalable anyway
2. Deliberately checking out an old commit in these repos seemed like the kind of thing we would never update, and we'd miss major releases because of it
3. This approach (using `date`) is extremely simple, and seems like it gives us basically what we want

I tested this manually by changing `etc/proto/Dockerfile:14` to `RUN date --date="2 weeks ago" +%s >/last_run_time` and confirming that `make proto` fails. I also tested the process for updating the docker image (by running `make proto` and confirming that the protobuf compiler was re-downloaded). Given that this is purely a convenience feature and internal-only, I didn't think it made sense to try and set up any kind of automated test.

Finally, I also ran an updated version of `make proto` and updated all of our generated `.pb.go` files

### CODE REVIEW

**Comment by msteffen** [comment]
Update: A few months ago, golang/protobuf made a change that caused it to generate backwards-incompatible go structs: https://github.com/golang/protobuf/issues/607

To get this change to build, I've also had to update our code to handle the new generated structs

File: etc/proto/run [hunk 1]
```diff
@@ -1,7 +1,23 @@
...
 
+# Make sure tools are all up-to-date so that our compiled protobufs aren't
+# generated by stale tools
+max_age="$(date --date="1 week ago" +%s)"
+if [[ "${max_age}" -gt "$(cat /last_run_time)" ]]; then
+  PRE="\e[1;31m~\e[0m"
+  echo -e "\e[1;31m~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\e[0m" >/dev/stderr
+  echo -e "${PRE} \e[1;31mWARNING:\e[0m pachyderm_proto is out of date" >/dev/stderr
+  echo -e "${PRE} please run" >/dev/stderr
+  echo -e "${PRE}   docker ps -a -q | xargs docker rm  "\#" rm all docker containers" >/dev/stderr
```
Comment (jdoliner) [inline etc/proto/run]: Why remove all containers? This would probably kill a kubernetes cluster that's running in containers.

Comment (msteffen) [reply etc/proto/run]: After `make proto` finishes running, the container that it was running in is still around, it's just stopped (and you need to delete the container to delete the `pachyderm_proto` image, and you need to delete the image for rebuilding the image to re-download the protobuf compiler).

I couldn't figure out a way to delete only the `pachyderm_proto` container, but if you know how to do that (maybe with `docker ps --filter`, which I briefly tried to figure out but couldn't?) I agree that that would be better

Comment (msteffen) [reply etc/proto/run]: Just figured it out, I think

File: etc/proto/run [hunk 2]
```diff
@@ -1,7 +1,23 @@
-#!/bin/sh
+#!/bin/bash
 
 tar xf /dev/stdin
 
+# Make sure tools are all up-to-date so that our compiled protobufs aren't
+# generated by stale tools
+max_age="$(date --date="1 week ago" +%s)"
```
[the remark this answers is not in the document]
Comment (msteffen) [reply etc/proto/run]: Done

**Pull request merged**